### PR TITLE
Added MeamFit code to pyiron_contrib

### PR DESCRIPTION
**Pull request by usaikia**
MeamFit was available only in pyiron_mpie. Here I coped the code the from pyiron_mpie and added into pyiron_contrib with a minor modification to copy vasprun.xml files from previously finished vasp md jobs.

### CODE REVIEW

**Comment by pmrv** [comment]
I'll be back next week and have a detailed look then, but it already looks fine.

**Comment by jan-janssen** [comment]
You still have to add the code to the `job_class_dict` https://github.com/pyiron/pyiron_contrib/blob/master/pyiron_contrib/__init__.py#L14

File: pyiron_contrib/atomistics/meamfit/meamfit.py [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+__maintainer__ = "Jan Janssen"
+__email__ = "janssen@mpie.de"
+__status__ = "development"
+__date__ = "Sep 1, 2017"
+
+
+class MeamFit(GenericJob):
+    def __init__(self, project, job_name):
+        """
+        ExampleJob generating a list of random numbers to simulate energy fluctuations.
```
Comment (pmrv) [inline pyiron_contrib/atomistics/meamfit/meamfit.py:28]: Needs to be updated.

File: pyiron_contrib/atomistics/meamfit/meamfit.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+    def structure(self, structure):
+        """
+
+        Args:
+            structure:
+
+        Returns:
+
+        """
+        raise ValueError('MeamFit has no structure!')
```
Comment (pmrv) [inline pyiron_contrib/atomistics/meamfit/meamfit.py:88]:
```suggestion
```

File: pyiron_contrib/atomistics/meamfit/meamfit.py [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+
+    def set_input_to_read_only(self):
+        """
+        This function enforces read-only mode for the input classes, but it has to be implement in the individual
+        classes.
+        """
+        self.input.read_only = True
+
+    def validate_ready_to_run(self):
+        pass
```
Comment (pmrv) [inline pyiron_contrib/atomistics/meamfit/meamfit.py:110]: Should check whether there's training calculations added already.

File: pyiron_contrib/atomistics/meamfit/meamfit.py [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+    def add_job_to_fitting(self, job_id, time_step_start=0, time_step_end=-1, time_step_delta=10, quantity='E0',
+                           weight=[1.0, 0.0, 0.0]):
+        """
+
+        Args:
+            job_id (int):
+            time_step_start (int): initial timestep - after equilibration
+            time_step_end (int): last timestep to use
+            time_step_delta (int):
+            quantity (str): ['E', 'F'] for fitting the energies use 'E' and 'F' for fitting the forces
```
Comment (pmrv) [inline pyiron_contrib/atomistics/meamfit/meamfit.py:136]: Why is the default not one of the allowed values?

File: pyiron_contrib/atomistics/meamfit/meamfit.py [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+            weight (list): default is one, but for lower temperatures use higher weights [100, 1000]
```
Comment (pmrv) [inline pyiron_contrib/atomistics/meamfit/meamfit.py:137]: Why is the default of length three?  What does it mean to use a weight >100-1000 if the default is 1?  A link to the official documentation would be enough.

File: pyiron_contrib/atomistics/meamfit/meamfit.py [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+                                   'Weights': [weight]})
+            self._calculation_dataframe = pd.concat([self._calculation_dataframe, df_tmp.set_index('Job id')])
+
+    # define routines that collect all output files
+    def collect_output(self):
+        """
+
+        Returns:
+
+        """
```
Comment (pmrv) [inline pyiron_contrib/atomistics/meamfit/meamfit.py:165]:
```suggestion
```

File: pyiron_contrib/atomistics/meamfit/meamfit.py [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+        self._potential_timings_dataframe = self._calculate_std(potential_timings_df,
+                                                                self._potential_performance_dataframe)
+        self.to_hdf()
+
+    def collect_logfiles(self):
+        """
+
+        Returns:
+
+        """
```
Comment (pmrv) [inline pyiron_contrib/atomistics/meamfit/meamfit.py:177]:
```suggestion
```

File: pyiron_contrib/atomistics/meamfit/meamfit.py [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+        pass
+
+    def from_directory(self, directory):
```
Comment (pmrv) [inline pyiron_contrib/atomistics/meamfit/meamfit.py:180]: Missing docstring.

File: pyiron_contrib/atomistics/meamfit/meamfit.py [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+        if not self.status.finished:
+            self.status.collect = True
+            self._import_directory = directory
+            self.input.read_input(posixpath.join(directory, 'settings'))
+            self._calculation_dataframe = self._read_calc_db(file_name="fitdbse", cwd=directory)
+            self.collect_output()
+            self.status.finished = True
+        else:
+            return
```
Comment (pmrv) [inline pyiron_contrib/atomistics/meamfit/meamfit.py:189]: Raising a `RuntimeError` here would be nicer.

File: pyiron_contrib/atomistics/meamfit/meamfit.py [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+    def to_hdf(self, hdf=None, group_name=None):
+        """
+
+        Args:
+            hdf:
+            group_name:
+
+        Returns:
+
+        """
```
Comment (pmrv) [inline pyiron_contrib/atomistics/meamfit/meamfit.py:201]:
```suggestion
```

File: pyiron_contrib/atomistics/meamfit/meamfit.py [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+    def from_hdf(self, hdf=None, group_name=None):
+        """
+
+        Args:
+            hdf:
+            group_name:
+
+        Returns:
+
+        """
```
Comment (pmrv) [inline pyiron_contrib/atomistics/meamfit/meamfit.py:220]:
```suggestion
```

File: pyiron_contrib/atomistics/meamfit/meamfit.py [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+            if 'File' in self._potential_timings_dataframe.columns:
+                self._potential_timings_dataframe = self._potential_timings_dataframe.set_index('File')
+
+    def _copy_vasprun_xml(self, cwd=None):
+        for job_id in self._calculation_dataframe.index:
+            self.project.load(job_id).decompress()
+            working_directory = self.project.get_job_working_directory(int(job_id))
+            shutil.copyfile(posixpath.join(working_directory, 'vasprun.xml'),
+                            posixpath.join(cwd, self._get_vasprun_name(job_id)))
+            self.project.load(job_id).compress()
```
Comment (pmrv) [inline pyiron_contrib/atomistics/meamfit/meamfit.py:238]: Move closer to `write_input` or even inline it.

File: pyiron_contrib/atomistics/meamfit/meamfit.py [hunk 13]
```diff
@@ -0,0 +1,356 @@
...
+
+    @staticmethod
+    def _write_calc_db(calculation_dataframe, file_name="fitdbse", cwd=None):
```
Comment (pmrv) [inline pyiron_contrib/atomistics/meamfit/meamfit.py:241]: Move closer to write_input or even inline it.

File: pyiron_contrib/atomistics/meamfit/meamfit.py [hunk 14]
```diff
@@ -0,0 +1,356 @@
...
+                             list(calculation_dataframe['Weights'])):
+                file, start_config, end_config, step, quantity, weight = entry
+                if isinstance(weight, list):
+                    weight = str(weight[0]) + ' ' + str(weight[1]) + ' ' + str(weight[2])
+                f.write(str(file) + ' ' + str(int(start_config)) + '-' + str(int(end_config)) + 's' + str(int(step))
+                        + ' ' + str(quantity) + ' ' + str(weight) + '\n')
+
+    @staticmethod
+    def _get_vasprun_name(job_id):
+        return 'vasprun_' + str(int(job_id)) + '.xml'
```
Comment (pmrv) [inline pyiron_contrib/atomistics/meamfit/meamfit.py:260]: Inline into `_copy_vasprun_xml`

File: pyiron_contrib/atomistics/meamfit/meamfit.py [hunk 15]
```diff
@@ -0,0 +1,356 @@
...
+        ExampleJob generating a list of random numbers to simulate energy fluctuations.
+
+        Args:
+            project: Project object (defines path where job will be created and stored)
+            job_name: name of the job (must be unique within this project path)
+        """
+        super(MeamFit, self).__init__(project, job_name)
+        self.__version__ = None
+        self.__name__ = "MeamFit"
+        self.input = MeamFitInput()
```
Comment (pmrv) [inline pyiron_contrib/atomistics/meamfit/meamfit.py:37]: Replace with a `DataContainer` with appropriate defaults.

**Review by pmrv — CHANGES_REQUESTED** [review]
Thanks @usaikia for moving it over, I think it'll be very useful!

Sorry for the nitty comments, but I'd like new code additions to be pretty good shape otherwise we'll never get around to it.

File: pyiron_contrib/atomistics/meamfit/meamfit.py [hunk 16]
```diff
@@ -0,0 +1,311 @@
...
+    def from_directory(self, directory):
+        if not self.status.finished:
+            self.status.collect = True
+            self._import_directory = directory
+            self.input.read_input(posixpath.join(directory, 'settings'))
+            self._calculation_dataframe = self._read_calc_db(file_name="fitdbse", cwd=directory)
+            self.collect_output()
+            self.status.finished = True
+        else:
+            return
```
Comment (pmrv) [inline pyiron_contrib/atomistics/meamfit/meamfit.py:209]:
```suggestion
            raise RuntimeError('Can import MEAMfit calculation into finished job.  Needs to be `initialized`.)
```

File: pyiron_contrib/atomistics/meamfit/meamfit.py (hunk 9 above)
Comment (pmrv) [reply pyiron_contrib/atomistics/meamfit/meamfit.py:209]: Check also if method is really necessary?

File: pyiron_contrib/atomistics/meamfit/meamfit.py (hunk 8 above)
Comment (usaikia) [reply pyiron_contrib/atomistics/meamfit/meamfit.py:200]: Done.

File: pyiron_contrib/atomistics/meamfit/meamfit.py [hunk 17]
```diff
@@ -0,0 +1,317 @@
...
+    def set_input_to_read_only(self):
+        """
+        This function enforces read-only mode for the input classes, but it has to be implement in the individual
+        classes.
+        """
+        self.input.read_only = True
+
+    def validate_ready_to_run(self):
+        if self._calculation_dataframe.empty:
+            raise ValueError()
```
Comment (pmrv) [inline pyiron_contrib/atomistics/meamfit/meamfit.py:106]:
```suggestion
            raise ValueError("No training data added yet!")   
```

File: pyiron_contrib/atomistics/meamfit/meamfit.py [hunk 18]
```diff
@@ -0,0 +1,317 @@
...
+            time_step_delta (int): time step
+            quantity (str): the property in the vasprun file to be fit to, and can take the values ['E0', 'Free‐Energy', 'Force']
+                            ‘E0’: to the total energy (specifically the E0, sigma‐>0 value in the vasprun file);
+                            ‘Free‐Energy’: will fit to the free energy (the F value in the vasprun file);  
+                            ‘Force’: will fit to atomic forces.
+                             Note that only the first two letters are in fact read in by MEAMfit, so that it is sufficient to
+                             write ‘Fr’ or ‘Fo’ for the second and third cases respectively.  
+            weight (list): default is [1.0, 0.0, 0.0].
+            
+            More information on these parameters are available in the MeamFit user manual- https://www.scd.stfc.ac.uk/Pages/MEAMfit-v2.aspx
```
Comment (pmrv) [inline pyiron_contrib/atomistics/meamfit/meamfit.py:140]: I cannot find a manual under the given link.  Is it not public?

File: pyiron_contrib/atomistics/meamfit/meamfit.py [hunk 19]
```diff
@@ -0,0 +1,317 @@
...
+            self.input['SEED'] = random.randint(0, 100000)
+        self.input.write_file(file_name="settings", cwd=self.working_directory)
+        self._write_calc_db(calculation_dataframe=self._calculation_dataframe,
+                            file_name="fitdbse",
+                            cwd=self.working_directory)
+        self._copy_vasprun_xml(cwd=self.working_directory)
+
+    def add_job_to_fitting(self, job_id, time_step_start=0, time_step_end=-1, time_step_delta=10, quantity='E0',
+                           weight=[1.0, 0.0, 0.0]):
+        """
```
Comment (pmrv) [inline pyiron_contrib/atomistics/meamfit/meamfit.py:125]:
```suggestion
        """
        Add output of VASP jobs to training data.
```

File: pyiron_contrib/atomistics/meamfit/meamfit.py [hunk 20]
```diff
@@ -0,0 +1,317 @@
...
+                            ‘Force’: will fit to atomic forces.
+                             Note that only the first two letters are in fact read in by MEAMfit, so that it is sufficient to
+                             write ‘Fr’ or ‘Fo’ for the second and third cases respectively.  
+            weight (list): default is [1.0, 0.0, 0.0].
+            
+            More information on these parameters are available in the MeamFit user manual- https://www.scd.stfc.ac.uk/Pages/MEAMfit-v2.aspx
+
+        Returns:
+
+        """
```
Comment (pmrv) [inline pyiron_contrib/atomistics/meamfit/meamfit.py:144]:
```suggestion
        """
        job = self.project.load(job_id)
        if job.__name__ != "Vasp":
            raise ValueError("Training data must be from VASP jobs!")
```

File: pyiron_contrib/atomistics/meamfit/meamfit.py [hunk 21]
```diff
@@ -0,0 +1,317 @@
...
+                                   'Weights': [weight]})
+            self._calculation_dataframe = pd.concat([self._calculation_dataframe, df_tmp.set_index('Job id')])
+
+    def _copy_vasprun_xml(self, cwd=None):
+        for job_id in self._calculation_dataframe.index:
+            self.project.load(job_id).decompress()
+            working_directory = self.project.get_job_working_directory(int(job_id))
+            shutil.copyfile(posixpath.join(working_directory, 'vasprun.xml'),
+                            posixpath.join(cwd, 'vasprun_' + str(int(job_id)) + '.xml'))
+            self.project.load(job_id).compress()
```
Comment (pmrv) [inline pyiron_contrib/atomistics/meamfit/meamfit.py:168]:
```suggestion
            job = self.project.load(job_id)
            job.decompress()
            working_directory = self.project.get_job_working_directory(int(job_id))
            shutil.copyfile(posixpath.join(working_directory, 'vasprun.xml'),
                            posixpath.join(cwd, 'vasprun_' + str(int(job_id)) + '.xml'))
            job.compress()
```

File: pyiron_contrib/atomistics/meamfit/meamfit.py [hunk 22]
```diff
@@ -0,0 +1,317 @@
...
+                            ‘Free‐Energy’: will fit to the free energy (the F value in the vasprun file);  
+                            ‘Force’: will fit to atomic forces.
+                             Note that only the first two letters are in fact read in by MEAMfit, so that it is sufficient to
+                             write ‘Fr’ or ‘Fo’ for the second and third cases respectively.  
+            weight (list): default is [1.0, 0.0, 0.0].
+            
+            More information on these parameters are available in the MeamFit user manual- https://www.scd.stfc.ac.uk/Pages/MEAMfit-v2.aspx
+
+        Returns:
+
```
Comment (pmrv) [inline pyiron_contrib/atomistics/meamfit/meamfit.py:143]:
```suggestion
        Raises:
            ValueError: if given job is a not a Vasp job
```

**Comment by pmrv** [comment]
I've made some small changes.  @usaikia Can you double check everything still works? Except for the manual link everything seems good to go for me.  If the manual is not available publically, just remove the link and merge.

File: pyiron_contrib/atomistics/meamfit/meamfit.py (hunk 18 above)
Comment (usaikia) [reply pyiron_contrib/atomistics/meamfit/meamfit.py:141]: The manual is not available publicly. In order to use the code users has to ask for permission individually from Andrew Ian Duff (andrew.duff@stfc.ac.uk)​​. Andrew send the code and manual to the user with permission to use the code.

Comment (pmrv) [reply pyiron_contrib/atomistics/meamfit/meamfit.py:141]: Ok, then let's just put a comment to that effect into the main docstring of the class.

**Comment by pmrv** [comment]
There should also be a reference to the MEAMFit paper: https://www.sciencedirect.com/science/article/pii/S0010465515001964#f000020
Here's an example of how to do that: https://github.com/pyiron/pyiron_contrib/blob/9d72f2ed5fe7d3e2f3ab000ef7602711e28356bc/pyiron_contrib/atomistics/pacemaker/job.py#L108

**Pull request merged**